feature_repository.go

```diff
@@ -0,0 +1,497 @@
+	if first {
+		// We were the first request to come in, so perform the API
+		// request, and...
+		apiResponse = doFetchRequest(gb)
+
+		// ...retrieve a list of channels to other goroutines requesting
+		// features for the same repository key, clearing the outstanding
+		// requests slot for this repository key...
+		outstandingRequestMutex.Lock()
+		chans := outstandingRequest[key]
+		delete(outstandingRequest, key)
+		outstandingRequestMutex.Unlock()
+
+		// ...then send the API response to all the waiting goroutines. We
+		// check that our channel is still in the list, in case the cache
+		// and the outstanding requests information has been cleared while
+		// we were making the request.
+		selfFound := false
+		for _, ch := range chans {
+			if ch != myChan {
+				ch <- apiResponse
+			} else {
+				// Don't send to ourselves, but record that our channel is
+				// still in the list.
+				selfFound = true
+			}
+		}
+
+		// Finally call the new feature data callback (from a single
+		// goroutine), assuming that the outstanding requests list hasn't
+		// been cleared in the meantime.
+		if apiResponse != nil && selfFound {
+			onNewFeatureData(key, apiResponse)
+			refresh.runBackgroundRefresh(gb)
+		}
+	} else {
+		// We were a later request, so just wait for the result from the
+		// goroutine performing the request on our channel.
+		apiResponse = <-myChan
+	}
+
+	// If something went wrong, we return an empty response, rather than
+	// nil.
+	if apiResponse == nil {
+		apiResponse = &FeatureAPIResponse{}
+	}
+	return apiResponse
+}
+
+func fetchFeaturesWithTimeout(gb *GrowthBook, timeout time.Duration) *FeatureAPIResponse {
+	if timeout == 0 {
+		return fetchFeatures(gb)
+	}
+	ch := make(chan *FeatureAPIResponse, 1)
+	timer := time.NewTimer(timeout)
+	go func() {
+		ch <- fetchFeatures(gb)
+	}()
+	select {
+	case result := <-ch:
+		return result
+	case <-timer.C:
+		return nil
+	}
+}
+
+func onNewFeatureData(key repositoryKey, data *FeatureAPIResponse) {
+	// If contents haven't changed, ignore the update, extend the stale TTL
+	version := data.DateUpdated
+	now := time.Now()
+	staleAt := now.Add(cacheStaleTTL)
+	existing := cache.get(key)
+	if existing != nil && version != "" && existing.version == version {
+		existing.staleAt = staleAt
+		return
+	}
+
+	// Update in-memory cache
+	cache.set(key, &cacheEntry{data, version, staleAt})
+
+	// Update features for all subscribed GrowthBook instances.
+	for _, inner := range refresh.instances(key) {
+		refreshInstance(inner, data)
+	}
+}
+
+// -----------------------------------------------------------------------------
+//
+//  AUTO-REFRESH PROCESSING
+
+// We store *only* the inner data structure of GrowthBook instances
+// here, so that the finalizer added to the main (outer) GrowthBook
+// instances will run, triggering an unsubscribe, allowing us to
+// remove the inner data structure here.
+type gbDataSet map[*growthBookData]bool
+
+type refreshData struct {
+	sync.RWMutex
```

Comment:
My understanding with mutexes is we can risk getting into a deadlock if the application panics before an unlock is performed, but that you can `defer` the unlock with RWMutex like you're doing, and so that may not be a concern. So I'm guessing this is pretty low risk but I just wanted to gather your thoughts on the potential risks here if it's possible that we can lock up users' apps in the event something goes terribly wrong here.

Reply:
Go's `defer` has semantics specifically designed to deal with this (details [here](https://go.dev/ref/spec#Defer_statements)). In short, `defer` statements are executed however the thread of execution leaves the function containing the `defer`, whether that's through a normal return, or because of a panic.

I will add some comments about locking strategies though. I try to follow a very stereotyped approach to these things where it's possible, so that it's easy to see that there is no opportunity for deadlock. In particular, in this code, every lock is paired with a deferred unlock, and mutexes are considered as belonging to the data structures that they protect, so are only ever locked within methods associated with those data structures. I believe that the way that that works means that it shouldn't be possible to get into a deadlock. (Actually, there is one place where I didn't follow the struct lock/deferred unlock approach, but I'll change that because it's not completely correct. Even though there's close to zero chance of a panic within the relevant critical sections, there's no reason not to refactor to remove even that possibility.)

Reply:
I'll add some notes about this in the documentation PR.